cumulus/pallets/parachain-system/src/parachain_inherent.rs

```diff
@@ -0,0 +1,552 @@
+///
+/// This structure contains a `sent_at` field and a reverse index. Using this information, we can
+/// identify a message inside a sorted collection by walking back `reverse_idx` positions starting
+/// from the last message that has the provided `sent_at`.
+///
+/// We use a reverse index instead of a normal index because sometimes the messages at the
+/// beginning of the collection are being pruned.
+///
+/// # Example
+///
+///
+/// For the collection
+/// `msgs = [{sent_at: 0}, {sent_at: 1}, {sent_at: 1}, {sent_at: 1}, {sent_at: 1}, {sent_at: 3}]`
+///
+/// `InboundMessageId {sent_at: 1, reverse_idx: 0}` points to `msgs[4]`
+/// `InboundMessageId {sent_at: 1, reverse_idx: 3}` points to `msgs[1]`
+/// `InboundMessageId {sent_at: 1, reverse_idx: 4}` points to `msgs[0]`
+#[derive(
+	Encode,
+	Decode,
+	DecodeWithMemTracking,
+	Clone,
+	Default,
+	sp_runtime::RuntimeDebug,
+	PartialEq,
+	TypeInfo,
+)]
+pub struct InboundMessageId {
+	/// The block number at which this message was added to the message passing queue
+	/// on the relay chain.
+	pub sent_at: BlockNumber,
+	/// The reverse index of the message in the collection of messages sent at `sent_at`.
+	pub reverse_idx: u32,
+}
+
+/// A message that was received by the parachain.
+pub trait InboundMessage {
+	/// The corresponding compressed message.
+	/// This should be an equivalent message that stores the same metadata as the current message,
+	/// but stores only a hash of the message data.
+	type CompressedMessage: Debug;
+
+	/// Gets the message data.
+	fn data(&self) -> &[u8];
+
+	/// Gets the relay chain number where the current message was pushed to the corresponding
+	/// relay chain queue.
+	fn sent_at(&self) -> RelayChainBlockNumber;
+
+	/// Converts the current message into a `CompressedMessage`
+	fn to_compressed(&self) -> Self::CompressedMessage;
+}
+
+/// A collection of inbound messages.
+#[derive(
+	codec::Encode,
+	codec::Decode,
+	codec::DecodeWithMemTracking,
+	sp_core::RuntimeDebug,
+	Clone,
+	PartialEq,
+	TypeInfo,
+)]
+pub struct InboundMessagesCollection<Message: InboundMessage> {
+	messages: Vec<Message>,
+}
+
+impl<Message: InboundMessage> InboundMessagesCollection<Message> {
+	/// Creates a new instance of `InboundMessagesCollection` that contains the provided `messages`.
+	pub fn new(messages: Vec<Message>) -> Self {
+		Self { messages }
+	}
+
+	/// Drop all the messages up to `last_processed_msg`.
+	pub fn drop_processed_messages(&mut self, last_processed_msg: &InboundMessageId) {
+		let mut last_processed_msg_idx = None;
+		let messages = &mut self.messages;
+		for (idx, message) in messages.iter().enumerate().rev() {
+			let sent_at = message.sent_at();
+			if sent_at == last_processed_msg.sent_at {
+				last_processed_msg_idx = idx.checked_sub(last_processed_msg.reverse_idx as usize);
+				break;
+			}
+			if sent_at < last_processed_msg.sent_at {
+				last_processed_msg_idx = Some(idx);
+				break;
+			}
```

Comment:
```suggestion
```

This is really not possible, as said before. The hash comparison on chain would already have failed before.

Reply:
The issue for adding a zombienet test: https://github.com/paritytech/polkadot-sdk/issues/9022

Reply:
As discussed offline this is not caused by what I described initially. It is because we may build on the same parent twice and we may receive the same messages again while `last_processed_msg` was updated. Added a comment.